architecture/2025-01-serverless-feature-spec.md

```diff
@@ -0,0 +1,391 @@
+      ports: {
+        web: {
+          containerPort: 3000
+        }
+      }
+    }
+    runtimes: {
++     aci: {
++       // Add ACI-specific properties here to punch-through the Radius abstraction, e.g. sku, osType, etc.
++       sku: 'Confidential' // 'Standard', 'Dedicated', etc.
++     }
++     ecs: {
++       // Add AWS ECS-specific properties here to punch-through the Radius abstraction
++     }
++      }
++   // connections to other resources remains the same for serverless containers as they are for Kubernetes containers
++   connections: {
++     redis: {
++       source: db.id
++     }
++   }
+  }
+}
+
++ // Definitions for other resources that the container can connect to remain unchanged
+resource db 'Applications.Datastores/redisCaches@2023-10-01-preview' = {
+  name: 'db'
+  properties: {
+    application: application
+    environment: environment
+  }
+}
+```
+
+### Step 5: Define a Dapr sidecar for the serverless container and connect it to a Dapr resource
+
+> Note: This step is only applicable if the serverless compute platform has native support for Dapr.
+
+The user defines a Dapr sidecar in the container definition and connects it to a Dapr resource.
+
+```diff
+resource demo 'Applications.Core/containers@2023-10-01-preview' = {
+  name: 'demo'
+  properties: {
+    application: application
+    container: {
+      image: 'ghcr.io/radius-project/samples/demo:latest'
+      ports: {
+        web: {
+          containerPort: 3000
+        }
+      }
+    }
+    runtimes: {
++     aci: {
++       // Add ACI-specific properties here to punch-through the Radius abstraction, e.g. sku, osType, etc.
++       sku: 'Confidential' // 'Standard', 'Dedicated', etc.
++     }
++     ecs: {
++       // Add AWS ECS-specific properties here to punch-through the Radius abstraction
++     }
++   }
++   extensions: [
++     {
++       kind: 'daprSidecar'
++       appId: 'demo'
++       appPort: 3000
++     }
++   ]
++   // connections to other resources remains the same for serverless containers as they are for Kubernetes containers
++   connections: {
++     redis: {
++       source: daprStateStore.id
++     }
++   }
+  }
+}
+
++ // Definitions for Dapr resources that the container can connect to remain unchanged
+resource daprStateStore 'Applications.Dapr/stateStores@2023-10-01-preview' = {
+  name: 'statestore'
+  properties: {
+    environment: environment
+    application: application
+  }
+}
+```
+
+### Step 6: Deploy the application to the serverless platform
+
+The user deploys the application to the serverless platform by running the `rad run` or `rad deploy` command with the application definition file targeting the serverless environment (e.g. `app.bicep`).
+
+### Step 7: View the serverless containers in the CLI and Radius Dashboard
+
+After successful deployment, the user can view the serverless containers in the CLI using the `rad app graph` command or via the Application Graph in the Radius Dashboard.
+
+## Key investments
```

Comment:
Should we add a key investment for Radius managing external environments? This could be a separately deployable feature that has value on its own, but it is definitely required for ACI support.

Reply:
I added a note in the risks section (see **Risk: Deploying to multiple clusters** ) where I call out that we might have to partially implement (or at least take into consideration) the [Inter-cluster app deployment and management](https://github.com/radius-project/roadmap/issues/42) feature. I hesitate to add this to the explicit set of features we plan to deliver as that itself is a separate (and possibly large) project that we probably don't want to fully implement as a subproject of serverless. In other words, we should only build enough functionality for Radius managing external environments that will satisfy our Serverless features. Let me know if you disagree and we can discuss further.